hspec-core/src/Test/Hspec/Core/Runner/Eval.hs

```diff
@@ -53,22 +53,27 @@ every seconds action = do
 
 type FormatResult = Either E.SomeException Result -> FormatM ()
 
-parallelize :: Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-parallelize isParallelizable e
-  | isParallelizable = runParallel e
-  | otherwise = runSequentially e
+parallelize :: Maybe QSem -> Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+parallelize jobsSem isParallelizable e
+  | isParallelizable = runParallel jobsSem e
+  | otherwise = runSequentially jobsSem e
 
-runSequentially :: (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-runSequentially e reportProgress formatResult = return $ do
-  result <- liftIO $ evalExample (e reportProgress)
+limitParallelism :: Maybe QSem -> IO a -> IO a
+limitParallelism jobsSem = case jobsSem of
+  Nothing -> id
+  Just s -> E.bracket_ (waitQSem s) (signalQSem s)
+
+runSequentially :: Maybe QSem -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+runSequentially jobsSem e reportProgress formatResult = return $ do
+  result <- liftIO $ limitParallelism jobsSem $ evalExample (e reportProgress)
```

Comment:
Perhaps the right thing to do would be to reserve one job for the sequential operations by starting the semaphore with one less than the provided number, then "give it back" by signaling the semaphore once when the final sequential job is done.  I considered doing it this way, but I couldn't see where in the code to add that final signalQSem operation.  If you can point that out to me, I'll be happy to change the implementation to do it that way.